### ISSUE TITLE
module aufile closes connection after playing wave file

### ISSUE THREAD

**Issue by hallerbachj** [alfredh/baresip#539]
Hi Alfred,

when I make a call using the /d parameter, wait until the connection is established, then change the audio source with /ausrc aufile, test.wav the wave file is beeing played, but after that the connection is closed. Is it possible, to not close the connection to give the user the chance to send for instance dtmf codes and close the connection if the right code is entered? Is it possible to play more than one wave file one after one by repeating the /ausrc command? I think, these functions maybe useful for other users too.

Are there some examples, how to get dtmf codes and make use of them?

Thanks

**Comment by alfredh** [comment]
Hi @hallerbachj 

this code in aufile module is calling error handler when the file is done:

```c
	/* check if audio buffer is empty */
	if (aubuf_cur_size(st->aubuf) < (2 * st->sampc)) {

		info("aufile: end of file\n");

		/* error handler must be called from re_main thread */
		if (st->errh)
			st->errh(0, "end of file", st->arg);
	}
```

furthermore, in the audio.c this handler is called:

```c
static void ausrc_error_handler(int err, const char *str, void *arg)
{
	struct audio *a = arg;
	MAGIC_CHECK(a);

	if (a->errh)
		a->errh(err, str, a->arg);
}
```

and finally in call.c will disconnect the call:

```
static void audio_error_handler(int err, const char *str, void *arg)
{
	struct call *call = arg;
	MAGIC_CHECK(call);

	if (err) {
		warning("call: audio device error: %m (%s)\n", err, str);
	}

	call_stream_stop(call);
	call_event_handler(call, CALL_EVENT_CLOSED, str);
}
```


the intended scenario was that the active call is playing a wav file
back to the remote peer, and when the file is done the call is disconnected.

if you prefer a different logic, we can discuss that

one work around for you is to use a WAV file with your audio and then
append a lot of silence to it :)

**Comment by hallerbachj** [comment]
Hi Alfred,

thanks for your reply. After commenting out 

> 	call_stream_stop(call);
> 	call_event_handler(call, CALL_EVENT_CLOSED, str);
the behaviour is exactly like I need it. I have tested it in the following way:

Make a call with the „d“ command
If the call ist accepted, switch the audio source with /ausrc aufile,“test.wav“ -> after playing is finished, the call is not terminated -> listening to dtmf-codes -> play the wave file again with /ausrc, „test.wav“ -> end the call with the „b“ command.

If I don’t comment out call_stream_stop(call) it is not possible to play the wave file again. Maybe you can add a parameter „AutoEndCall“ 0=hold the call after playing the wave file, 1=disconnect after playing? Or maybe we should introduce a flag CALL_EVENT_HOLD?
I think for some kind of answering machine („press 1 if you want this and 2 for that“) this is also a good solution. What do you think about it?

**Comment by hallerbachj** [comment]
Hi Alfred,

thank you, this is really a good and flexible solution!

I’m looking forward to testing it;-)

**Comment by alfredh** [comment]
hi @hallerbachj 

I made a suggested PR for this audio error event.

could you please review/test it ?

I am not 100% sure about the solution yet. perhaps AUDIO_EVENT is better ?


Alfred

**Comment by alfredh** [comment]
in your case it is probably better to clone the menu module and make your own
local custom changes, such as handling the audio error event and playing dtmf tones etc.

the menu.so module has a basic SIP client behaviour, which is supposed to work
for most people.

there are other modules such as echo.so and b2bua.so that replace menu.so
but have a completely different logic/behaviour.

**Comment by alfredh** [comment]
fixed in commit 7b6b8a05ebf44d566810b3a60b8799b6f8da5ddd
